WordPress/gutenberg
# Block Library: Add width attribute for resizable Column blocks

This pull request seeks to implement an initial exploration of resizable column blocks. It should serve as some foundation upon which to improve the usefulness of the Columns block.

![columns](https://user-images.githubusercontent.com/1779930/57376626-7ed8f600-716e-11e9-97fa-c1634a92cb9e.gif)

In this first iteration:

- An empty Column block displays with the "Button" appender introduced in #14241, similar to the Group block (#14943). The thinking is that this allows for a simpler gauge of columns width before adding content.
- An individual column block can be assigned with a `width` attribute which overrides the default equal distribution of column widths in a Columns block. Widths of all adjacent Column blocks are updated automatically, and widths redistributed when a Column is added or removed from the Columns block.

Not yet included, but planned for future iterations:

- A draggable resizer handle to adjust columns width. I have the bulk of the code for this complete already.
- Quick-select "template" options for columns distribution when initially creating a Columns block.
- Improved methods to select a Column and a Columns block to adjust settings.

**Specific Behavior:**

- When changing the width of a Column, the amount of increase or decrease is respectively removed or added in equal distribution across the Column blocks _after_ the block being changed, unless the Column block is the last in the set (in which case, the redistribution occurs in reverse).
- When adding or removing a Column to a Columns block, the new Column is assigned a width as if there were equal distribution. For example, increasing Columns from 2 to 3 would assume a width of 33.3% for the new block (100 / 3). This width (or, in the case of removal, the width of the removed block) is removed from or added to other blocks in the Columns.
- The default behavior of Columns remains unchanged in that `width` is only assigned or adjusted if a width value was assigned to a Column block in the set. This also helps ensure backwards-compatibility with previously-existing Column blocks.

**Testing Instructions:**

Verify that you can resize a Column block, and that in doing so, adjacent Column blocks are resized accordingly. Similarly, verify that adding or removing new Columns should redistribute widths to/from other columns.

1. Navigate to Posts > Add New
2. Insert a Columns block
3. Select the Column block (ArrowUp and ArrowDown are reliable here)
4. Adjust width value in the Inspector
5. Observe that the second column updates its width accordingly (to a sum of 100)
6. Select the Columns block
7. Increase / decrease Colujmsn count
8. Observe that the previous two columns update their widths accordingly (to a sum of 100)
9. Repeat steps 3 through 8 with a variation of Column combinations

**Remaining tasks:**

- [x] There is occasionally some issue in columns redistribution related to precision, where the sum total of Column widths may not equal 100 (but rather, e.g. 99.98%)
